lib/test-collection.js

```diff
@@ -195,18 +195,22 @@ class TestCollection extends EventEmitter {
 	}
 
 	build() {
-		const serialTests = new Sequence(this._buildTests(this.tests.serial), this.bail);
-		const concurrentTests = new Concurrent(this._buildTests(this.tests.concurrent), this.bail);
-		const allTests = new Sequence([serialTests, concurrentTests]);
+		const sequenceTests = context => {
+			const serialTests = new Sequence(this._buildTests(this.tests.serial, context), this.bail);
+			const concurrentTests = new Concurrent(this._buildTests(this.tests.concurrent, context), this.bail);
+			return new Sequence([serialTests, concurrentTests]);
+		};
 
 		let finalTests;
 		// Only run before and after hooks when there are unskipped tests
 		if (this._hasUnskippedTests()) {
```

Comment:
`_hasUnskippedTests()` depends on `allTests` having been created. Wrapping the test building in `sequenceTests()` won't work. Looks like this behavior isn't covered by an integration test though.

I think it's fine to still build `beforeHooks` and `afterHooks`. The trick is to not *run* them if all tests are skipped. We still "run" the skipped tests since we want to show them in the test results, but they don't actually execute.

Reply:
If I created the context above and injected it directly, I think it should have work as intended, but why did the `after` hook still got a mutated(mutations from `afterEach`, `beforeEach`, `tests`) context, when I ran the tests?

```javascript
const context = {context: {}};

const serialTests = new Sequence(this._buildTests(this.tests.serial, context), this.bail);
const concurrentTests = new Concurrent(this._buildTests(this.tests.concurrent, context), this.bail);
const allTests = new Sequence([serialTests, concurrentTests]);

...

if (this._hasUnskippedTests()) {
    const beforeHooks = new Sequence(this._buildHooks(this.hooks.before, null, context));
    const afterHooks = new Sequence(this._buildHooks(this.hooks.after, null, context));
    finalTests = new Sequence([beforeHooks, allTests, afterHooks], true);
} else {
    finalTests = new Sequence([allTests], true);
}
```